Reuse generator options used while creating app for updating Rails versions

prathamesh-sonpatki opened the pull request:
- Saves all the options passed while creating new app into `.rails`
  file.
- Reuses these options when `rake rails:update` is run.
- A sample `.rails` file - https://gist.github.com/prathamesh-sonpatki/dec79cf253ed7e5dc6ee

prathamesh-sonpatki commented:
r? @senny

senny commented:
@prathamesh-sonpatki could you squash the commits, add a changelog entry and also credit the original auther from #22042 in the commit message and changelog entry?

Also one important question we should answer is: What happens with apps that have been generated before this patch? Maybe output a message and make it possible to supply the flags on `rails:update` and then write the `.rails`. This would guarantee a smooth upgrade path.

prathamesh-sonpatki commented:
@senny sure I was waiting for feedback before adding changelog :).

Regarding existing apps, we can write the `.rails` file as you mentioned and helpful message when `rake rails:update` task run. Should I take care of it in a followup issue?

prathamesh-sonpatki committed:
Reuse generator options used while creating app for updating Rails versions

- Saves all the options passed while creating new app into `.rails`
  file.
- Reuses these options when `rake rails:update` is run.
- A sample `.rails` file -
  https://gist.github.com/prathamesh-sonpatki/dec79cf253ed7e5dc6ee
- Added required test cases for checking that `.rails` gets created with
  correct keys and gets used when `bin/rake rails:update` is used.
- Fixes https://github.com/rails/rails/pull/22042

senny commented:
@prathamesh-sonpatki Let's include the upgrade path in the same commit.

prathamesh-sonpatki commented:
@senny Squashed the commits and added changelog.

prathamesh-sonpatki commented:
Sure. I will add the upgrade path as well.

prathamesh-sonpatki commented:
@senny I did a quick check as follows -

```
be rake rails:update --skip-active-record
```

But it is failing as rake is not understanding the options. I think we have to convert this task to rails command infrastructure first before allowing users to pass the options.

kaspth commented:
Overall I think this is a good path to take, but I have some reservations.

Having `.rails` as a hidden file feels like a wart and not a bona fide convention. Since it works per project and can be thought of as configuring the `rails:update` command, we could add a `config/command_options/rails/update.yml` file instead. Would sharpen the purpose and clarify intent for the team member saddled with updating.

Second, wouldn't it make more sense to store only the passed in options?

senny commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -686,6 +686,19 @@ def test_after_bundle_callback
+  def test_dotrails_file_present
+    run_generator
+
+    assert_file ".rails"
```
We should also test the contents of the file. Let's run the generator with some options and see what ends up in the file.

senny replied:
We should also expand the test case to run `bin/rails rails:update` and see that skipped parts are not added again. This way we get full stack coverage.

prathamesh-sonpatki replied:
@senny Added missing tests.

senny commented:
@prathamesh-sonpatki I see. Ok let's hold off with the upgrade path and wait for @kaspth with progress on the new command infrastructure.

prathamesh-sonpatki commented:
@kaspth 

> we could add a config/command_options/rails/update.yml file instead. Would sharpen the purpose and clarify intent for the team member saddled with updating.

Agree. wdyt about `config/rails/update_options.yml` instead of `config/command_options/rails/update.yml`?

> Second, wouldn't it make more sense to store only the passed in options?

We are storing the options after they are parsed just to make sure that we are not storing any `incorrect` or `unsupported` option passed by user. Having complete list of options will make sure that the app will be updated using `same` options that were used to create the app.

claudiob commented:
Sorry for jumping into the discussion only now.

Could you explain why exactly this file is needed? Would it be impossible to infer the same information from (say) the modules included in the `config/application.rb` of your app or the gems in your `Gemfile`?

I typically run `rails new` skipping all the modules that I don't need right away, and then add them one by one when needed. I'm not sure the yml file would help in this case.

Maybe I'm just missing something… let me know :ribbon:

prathamesh-sonpatki commented:
@claudiob Good point. Yes it will be possible to infer this information from the `config/application.rb`, but why do that when we can be explicit about what options are going to be used for updating Rails version. Additionally it will be extra logic related to parsing files (based on regex or something) when we can mostly reuse the existing generator.

Other than that as you mentioned, some users can manually add the modules as and when required during the lifetime of the app. When they will want to upgrade the app, this file can actually guide them about which options were originally used while creating the app. We can show the list of options to the user before beginning with upgrade and add a prompt for allowing them to update the options as per their needs before continuing the upgrade.

prathamesh-sonpatki commented:
We will also have a place for adding new options and removing old options during a major upgrade where new modules are added.

prathamesh-sonpatki commented on railties/lib/rails/tasks/framework.rake:
```diff
@@ -45,13 +45,30 @@ namespace :rails do
+                    end
+          options.merge!(with_dispatchers: true)
+          options.merge!(force: true) if ENV['FORCE_RAILS_UPDATE']
+
```
Had to add this constant so that tests don't wait for user input for updating `secrets.yml` in tests.

senny commented:
A few thoughts from my point of view.

I think we should handle this similarly to `~/.railsrc`. It's a common use-case to have a dotfile that serves as a starting point when running commands. To me `bin/rails rails:update` should be thought of a command similar to `rails new`. You can pass it the same options to manipulate the outcome. Having the `.rails` file simply serves as a starting point. You could still overwrite these flags. We could implement it in a way that running `bin/rails rails:update` will be explicit about what happens. Something along the lines of:

```
$ bin/rails rails:update
using options from .rails
running with: --skip-spring --skip-turbolinks
proceed (y/n)?
```

This way it's clear where the options are coming from and the user has a chance to manipulate the file or rerun the command with different options.

Having a file in `config/` seems over the top. It's simply remembering how the app was generated. I don't se much use outside of that. Otherwise the format would need to be more complex and actually describe the components you are using. So that modifications also have effect on the Rails boot process.

Infering the options from `config/application.rb` seems overly complex with a lot of error potential. I'd rather do a simple thing that the user can control than trying to figure everything out. Remember that this is used only a handful of times when upgrading major and minor versions.

claudiob commented:
@senny :+1: If it's just a "nice plus" and you can still provide options on the command line, then it's a great idea! Thanks for the explanation.

prathamesh-sonpatki commented:
@kaspth your thoughts about this?

schneems commented:
This looks like a local config for only one app, after this is merged could we have a global config? I would LOVE to set my database to `postgres` for every `rails new` app.

rafaelfranca commented:
> I would LOVE to set my database to postgres for every rails new app.

It is already possible. Just place `-d postgresql` inside `~/railsrc`

kaspth commented:
Agree with @senny 👍

schneems commented:
@rafaelfranca Did not know. In that case, what about a `rails config:add -d postgresql` command or similar that adds things to that file?

rafaelfranca commented:
You usually only change that file once, what is the advantages of having a command to write it. I for example have it in my dotfiles and clone in every machine I use https://github.com/rafaelfranca/dotfiles/blob/master/dotfiles/railsrc#L1

schneems commented:
Discoverability, I've been contributing to Rails for years and didn't know it existed. Support for setting different dev environments like linux/windows since this won't work on windows:

```
$ echo "-d postgresql" >> ~/railsrc
```

rafaelfranca commented:
Looks good.

senny commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -686,6 +687,40 @@ def test_after_bundle_callback
+    run_generator [destination_root, "--skip-active-record", "--skip-action-cable"]
+    generator_options = YAML.load File.read(".rails")
+
+    assert_equal true, generator_options["skip_active_record"]
```
what would an assertion on the full generator_options look like?

```ruby
assert_equal {...}, generator_options
```

prathamesh-sonpatki replied:
@senny It would include all options stored in `.rails` file.

senny replied:
What options does it include in addition to the ones specified?

prathamesh-sonpatki commented:
@senny I am planning to change this implementation to store only those options that were added by user in *raw* format, instead of storing all options *after* they are parsed by us. 

The implementation can be similar to what we have for storing and parsing `.railsrc`.

We will also be able to add this prompt when `rails:update` is run - 

```
$ bin/rails rails:update
using options from .rails
running with: --skip-spring --skip-turbolinks
proceed (y/n)?
```
Does that sound good to you?

senny commented:
I'd welcome a format like `~/.railsrc`. It's simple but gets the job done. Any reason why we went with serializing the parsed options till now? Will we lose something?

prathamesh-sonpatki commented:
> Any reason why we went with serializing the parsed options till now?

I had done to avoid parsing the raw options again while running `rails:update` task.

> Will we lose something?
No. Because our generator code will always generate default values for missing options. So it won't be a problem.

senny commented:
@prathamesh-sonpatki great. Let's see how this approach turns out. Would be nice if .rails would just contain the options passed by the user. This should be easy enough to edit and commit later on. Also the prompt seems much nicer.

prathamesh-sonpatki commented:
@senny @rafaelfranca Please review.


Now we only store options passed by user in `.rails` and this prompt is shown to users when they try to update the app.

Additionally if you have a `.railsrc`, those options will also be stored in `.rails` file.

``` sh
prathamesh ~/Projects/reproductions/dotapp 2.3.0  -> 
$ cat .rails 
--skip-action-cable --dev
prathamesh ~/Projects/reproductions/dotapp 2.3.0  -> 
$ bin/rails app:update
Using options from .rails
Running with: --skip-action-cable --dev
Proceed?(y/n) 
```

matthewd commented:
I'm pretty unconvinced this is worth a top level file, hidden or not... particularly when it's only useful once a year or so. :neutral_face: 

Did we dismiss the idea of reverse-engineering the options from the current file contents?

senny commented:
@matthewd why do you think a file, toplevel or not, is such a big deal? From my point trying to guess what the user actually typed will be much more code and way more error prone. Why should we be smart about it when we can just remember what was typed in the first place?

matthewd commented:
A new top level file is something I must look past every time I enter any application directory... we currently generate 5 files, one of which I already consider unfortunate-but-justifiable. A 20% increase is in "extraordinary change requires extraordinary justification" territory to me.

Even inside `config/` we have a pretty short list of files, most of which are actively used/modified. Maybe we could hide it in a comment in `config/application.rb`?

---

The most obvious advantage of being smart (which for skipped components I think mostly just means looking at the `require` calls in `application.rb`) is that we can work based on what's currently active, rather than what was active when the generator was first run: I often forget to use the appropriate `--skip-*` option and remove a thing by hand, and I'm sure a more practiced skipper might choose to re-add a component later when they find they need it.

And as for `--api`, which I believe was the primary motivator to add this: it seems `config.api_only` would be a trivial and most-correct way to check that.

senny commented:
"Hiding" it in `config/application.rb` is fine for me as well. I can see the appeal of having us detect the flags based on the current state of the project. I too sometimes modify certain configs after generating the application. Having a comment that I can modify in advance would also solve part of that issue.

`--api` was not the main motivation for me. It happens a lot when running `bin/rake rails:update` that it adds back stuff that you didn't want to use in the first place `--skip-spring` comes to mind.

prathamesh-sonpatki commented:
We started this based on https://github.com/rails/rails/pull/22042 and some discussion in basecamp about rails command line infrastructure.

prathamesh-sonpatki commented:
@matthewd @senny If we detect the flash based on current state of project, we still want to give users an option to do an upgrade without those detected flags right?  Or are you suggesting that the logic to detect the components should be internal part of the upgrade process?

> Maybe we could hide it in a comment in config/application.rb?

A comment which can be read and parsed during update process?

rmacklin commented:
Should this PR be closed because https://github.com/rails/rails/pull/29645 has been merged?
